source/grid/grid_in.cc

```diff
@@ -2811,6 +2815,420 @@ GridIn<dim, spacedim>::read_assimp(const std::string &filename,
 #endif
 }
 
+namespace
+{
+  /**
+   * Find the reference cell type from the number of vertices.
+   */
+  template <int dim>
+  ReferenceCell::Type
+  n_vertices_to_type(const std::size_t n_vertices)
+  {
+    switch (dim)
+      {
+        case 1:
+          Assert(n_vertices == 2, ExcNotImplemented());
+          return ReferenceCell::Type::Line;
+        case 2:
+          switch (n_vertices)
+            {
+              case 3:
+                return ReferenceCell::Type::Tri;
+              case 4:
+                return ReferenceCell::Type::Quad;
+              default:
+                Assert(false, ExcNotImplemented());
+            }
+          break;
+        case 3:
+          {
+            switch (n_vertices)
+              {
+                case 4:
+                  return ReferenceCell::Type::Tet;
+                case 5:
+                  return ReferenceCell::Type::Pyramid;
+                case 6:
+                  return ReferenceCell::Type::Wedge;
+                case 8:
+                  return ReferenceCell::Type::Hex;
+              }
+            break;
+          }
+        default:
+          Assert(0 < dim && dim <= 3, ExcNotImplemented())
+      }
+    return ReferenceCell::Type::Invalid;
+  }
+
+  /**
+   * Convert between exodus and deal face numbers.
+   *
+   * This is templated so that the compiler can discard the irrelevant case
+   * statements.
+   */
+  template <int dim>
+  unsigned int
+  exodus_face_to_deal_face(const ReferenceCell::Type type,
+                           const unsigned int        face_n)
+  {
+    switch (dim)
+      {
+        case 1:
+          Assert(false, ExcNotImplemented());
+          break;
+        case 2:
+          switch (type)
+            {
+              case ReferenceCell::Type::Tri:
+                AssertIndexRange(face_n, 3);
+                // no table needed
+                return face_n;
+              case ReferenceCell::Type::Quad:
+                {
+                  AssertIndexRange(face_n, 4);
+                  constexpr std::array<unsigned int, 4> exodus_to_deal{
+                    {2, 1, 0, 3}};
```

Comment:
Could you provide a link/reference so that I can double check these values.

Reply:
It's on page 28 of this document:
http://gsjaardema.github.io/seacas/exodusII-new.pdf